### PR TITLE
Zoom on cursor | Issue #940

### PR DESCRIPTION
**Pull request by Jnforja**
This pull request is a first attempt at solving issue #940 

Although I checked for regression bugs, some might have escaped me. So I'd appreciate some help checking for those :)

Also, `viewportCoordsToSceneCoords()` and `sceneCoordsToViewportCoords()` have unused parameters. I want to fix that but haven't done so because I'm not sure if those parameters will be necessary to fix some regression bug that might have escaped me. So after we've checked for regressions I'd like to remove those unused parameters.

Edit: If I understood correctly, this PR also fixes issue #2163

### CODE REVIEW

**Comment by dwelle** [comment]
Completely forgot about this PR. Will review tomorrow, but great work from the looks of it!

**Review by lipis — APPROVED** [review]

File: src/appState.ts [hunk 1]
```diff
@@ -53,7 +53,10 @@ export const getDefaultAppState = (): Omit<
     isResizing: false,
     isRotating: false,
     selectionElement: null,
-    zoom: 1,
+    zoom: {
+      value: 1,
+      translation: { x: 0, y: 0 },
+    },
```
Comment (dwelle) [inline src/appState.ts:59]: we'll need a migration for this in `restore.ts`

**Review by dwelle — CHANGES_REQUESTED** [review]

**Comment by dwelle** [comment]
> Edit: If I understood correctly, this PR also fixes issue #2163

Looks like it :).

File: src/appState.ts (hunk 1 above)
Comment (Jnforja) [reply src/appState.ts:59]: I'll take a look at it 👍

Comment (Jnforja) [reply src/appState.ts:59]: Should be taken care of now. Tested locally, and it worked fine.

**Comment by dwelle** [comment]
@Jnforja pushed some refactor + changed [`canvasZoom` to be `zoom.value`](https://github.com/excalidraw/excalidraw/pull/2319/commits/bb3d6c237d508dfd839f35f606622a7d67db7d3e) so that the identity check works. Lemme know if you spot any issues.

**Comment by Jnforja** [comment]
@dwelle Nice one noticing that issue with `canvasZoom`. Thanks for letting me know.

I checked the refactor and all seems fine.

Can I assume that you also didn't find more regression bugs and remove those unused parameters from `viewportCoordsToSceneCoords()` and `sceneCoordsToViewportCoords()`?

**Comment by dwelle** [comment]
> Can I assume that you also didn't find more regression bugs and remove those unused parameters from `viewportCoordsToSceneCoords()` and `sceneCoordsToViewportCoords()`?

Can't guarantee as I haven't fully reviewed it (there's a lot of code), but if there are regression we'll know soon enough and fix them later.

What unused params are you talking about?

**Comment by Jnforja** [comment]
> What unused params are you talking about?

`canvas` and `scale`. They used to be necessary due to `getZoomOrigin()` being called in `viewportCoordsToSceneCoords()` and `sceneCoordsToViewportCoords()`. But that's no longer true.

What I'm not sure about is what role `scale` used to play. So I'm not 100% sure I shouldn't be taking it into consideration.

**Comment by dwelle** [comment]
Oh, right. I was checking the code in `master` branch so I didn't see it :). Can't say, I haven't implemented the original zoom behavior so can't say. Let's go ahead.

**Comment by Jnforja** [comment]
Ok! I'll go ahead and remove them :)

**Comment by Jnforja** [comment]
It's done

**Comment by lipis** [comment]
Removing feels nice! Let's merge it.

**Comment by lipis** [comment]
Thanks @Jnforja! ❤️

**Pull request merged**